**In this notebook we will note the diffrences between the accuracy of Conv2d and Separableconv2D**

my english is weak so plz manage 
In English 
Here , first we will import libraries than assign the path of directories where over data is present than flow data from directories , scale the the pixel using **Image_data_generator** class

In urdu
yahan per hmm na libraries ko import kia h 

In [ ]:

import os
from keras_preprocessing import image
from keras_preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras import layers
from keras import optimizers

In [69]:
train_dir = '/content/drive/My Drive/b and a/train'
val_dir  = '/content/drive/My Drive/b and a/valid'
test_dir = '/content/drive/My Drive/b and a/test'

In [70]:

train_gen = ImageDataGenerator(rescale= 1./255)
val_gen = ImageDataGenerator(rescale=1./255)
test_gen = ImageDataGenerator(rescale= 1./255)


In [71]:
train_data_gen = train_gen.flow_from_directory(train_dir,class_mode="binary",target_size=(224,224))
val_data_gen = val_gen.flow_from_directory(val_dir,class_mode="binary",target_size=(224,224))





Found 4000 images belonging to 2 classes.
Found 300 images belonging to 2 classes.


In [30]:
print(img)

[0. 0. 0. 1. 1. 0. 0. 0. 1. 1. 1. 0. 0. 1. 1. 1. 0. 1. 0. 0. 0. 0. 0. 1.
 0. 1. 0. 0. 0. 1. 1. 0.]


In [ ]:
for batch_input , batch_label in train_data_gen:
  print('batch_input  ', batch_input.shape )
  print('batch_label  ', batch_label.shape )
  break

batch_input   (32, 224, 224, 3)
batch_label   (32,)


Here we will create a model using one conv2d , two layers of Maxpooling2D ,one layer of Flatten and two layers of Dense 

In [73]:
model_1 = Sequential()
model_1.add(layers.Conv2D(32,(3,3) ,activation = 'relu' ,input_shape= (224,224,3)))
model_1.add(layers.MaxPooling2D( (2,2) ))
model_1.add(layers.MaxPooling2D( (2,2) ))
model_1.add(layers.Flatten())
model_1.add(layers.Dense(16,activation= "relu"))
model_1.add(layers.Dense(1 , activation = 'sigmoid' ))

model_1.compile(optimizer=optimizers.RMSprop(2e-5),loss="binary_crossentropy", metrics=["accuracy"])
model_1.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 222, 222, 32)      896       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 111, 111, 32)      0         
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 55, 55, 32)        0         
_________________________________________________________________
flatten (Flatten)            (None, 96800)             0         
_________________________________________________________________
dense (Dense)                (None, 16)                1548816   
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 1,549,729
Trainable params: 1,549,729
Non-trainable params: 0
______________________________________________

we can see that by using the above architecture of model_1 our validation accuracy has gone to 78 in 20 epochs and minimum loss is 50.66 


In [75]:
history_1 = model_1.fit_generator(train_data_gen,steps_per_epoch=40,epochs=10,validation_data=val_data_gen,validation_steps=10)


Instructions for updating:
Please use Model.fit, which supports generators.
Epoch 1/10
40/40 [==============================] - 979s 24s/step - loss: 0.6841 - accuracy: 0.5453 - val_loss: 0.6688 - val_accuracy: 0.6033
Epoch 2/10
40/40 [==============================] - 501s 13s/step - loss: 0.6632 - accuracy: 0.6008 - val_loss: 0.6495 - val_accuracy: 0.6800
Epoch 3/10
40/40 [==============================] - 337s 8s/step - loss: 0.6421 - accuracy: 0.6664 - val_loss: 0.6296 - val_accuracy: 0.7067
Epoch 4/10
40/40 [==============================] - 249s 6s/step - loss: 0.6228 - accuracy: 0.6609 - val_loss: 0.6150 - val_accuracy: 0.7033
Epoch 5/10
40/40 [==============================] - 152s 4s/step - loss: 0.6032 - accuracy: 0.6945 - val_loss: 0.6025 - val_accuracy: 0.7300
Epoch 6/10
40/40 [==============================] - 109s 3s/step - loss: 0.5939 - accuracy: 0.7063 - val_loss: 0.5898 - val_accuracy: 0.7400
Epoch 7/10
40/40 [==============================] - 82s 2s/step - loss: 0.59

In [80]:
history_1a = model_1.fit_generator(train_data_gen,steps_per_epoch=40,epochs=10,validation_data=val_data_gen,validation_steps=10)

Epoch 1/10
40/40 [==============================] - 4s 105ms/step - loss: 0.5592 - accuracy: 0.7336 - val_loss: 0.5465 - val_accuracy: 0.7633
Epoch 2/10
40/40 [==============================] - 4s 103ms/step - loss: 0.5525 - accuracy: 0.7297 - val_loss: 0.5403 - val_accuracy: 0.7633
Epoch 3/10
40/40 [==============================] - 4s 103ms/step - loss: 0.5425 - accuracy: 0.7500 - val_loss: 0.5350 - val_accuracy: 0.7700
Epoch 4/10
40/40 [==============================] - 4s 103ms/step - loss: 0.5355 - accuracy: 0.7531 - val_loss: 0.5264 - val_accuracy: 0.7733
Epoch 5/10
40/40 [==============================] - 4s 104ms/step - loss: 0.5256 - accuracy: 0.7500 - val_loss: 0.5236 - val_accuracy: 0.7633
Epoch 6/10
40/40 [==============================] - 4s 103ms/step - loss: 0.5060 - accuracy: 0.7680 - val_loss: 0.5610 - val_accuracy: 0.6967
Epoch 7/10
40/40 [==============================] - 4s 104ms/step - loss: 0.5131 - accuracy: 0.7680 - val_loss: 0.5226 - val_accuracy: 0.7633
Epoch 

In [74]:
# same architecture as above (model 1 ) but just change conv2d to separable conv2d 
from keras.models import Sequential
from keras import layers
from keras import optimizers

model_2 = Sequential()
model_2.add(layers.SeparableConv2D(32,(3,3) ,activation = 'relu' ,input_shape= (224,224,3)))
model_2.add(layers.MaxPooling2D( (2,2) ))
model_2.add(layers.MaxPooling2D( (2,2) ))
model_2.add(layers.Flatten())
model_2.add(layers.Dense(16,activation= "relu"))
model_2.add(layers.Dense(1 , activation = 'sigmoid' ))

model_2.compile(optimizer=optimizers.RMSprop(2e-5),loss="binary_crossentropy", metrics=["accuracy"])
model_2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
separable_conv2d (SeparableC (None, 222, 222, 32)      155       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 111, 111, 32)      0         
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 55, 55, 32)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 96800)             0         
_________________________________________________________________
dense_2 (Dense)              (None, 16)                1548816   
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 17        
Total params: 1,548,988
Trainable params: 1,548,988
Non-trainable params: 0
____________________________________________

we can see that by using the above architecture of model_2 our validation accuracy has gone to 76.33 in 20 epochs and minimum loss on 20 epochs is 51.46


In [76]:
history_2 = model_2.fit_generator(train_data_gen,steps_per_epoch=40,epochs=10,validation_data=val_data_gen,validation_steps=10)

Epoch 1/10
40/40 [==============================] - 17s 419ms/step - loss: 0.6855 - accuracy: 0.5406 - val_loss: 0.6644 - val_accuracy: 0.6567
Epoch 2/10
40/40 [==============================] - 16s 400ms/step - loss: 0.6604 - accuracy: 0.6094 - val_loss: 0.6461 - val_accuracy: 0.6733
Epoch 3/10
40/40 [==============================] - 8s 210ms/step - loss: 0.6412 - accuracy: 0.6609 - val_loss: 0.6294 - val_accuracy: 0.7133
Epoch 4/10
40/40 [==============================] - 9s 225ms/step - loss: 0.6221 - accuracy: 0.6742 - val_loss: 0.6154 - val_accuracy: 0.6900
Epoch 5/10
40/40 [==============================] - 9s 237ms/step - loss: 0.6142 - accuracy: 0.6805 - val_loss: 0.6040 - val_accuracy: 0.7100
Epoch 6/10
40/40 [==============================] - 5s 129ms/step - loss: 0.5916 - accuracy: 0.7063 - val_loss: 0.6294 - val_accuracy: 0.5867
Epoch 7/10
40/40 [==============================] - 5s 129ms/step - loss: 0.5907 - accuracy: 0.7000 - val_loss: 0.5860 - val_accuracy: 0.7167
Epoc

In [79]:
history_2a =  model_2.fit_generator(train_data_gen,steps_per_epoch=40,epochs=10,validation_data=val_data_gen,validation_steps=10)

Epoch 1/10
40/40 [==============================] - 5s 118ms/step - loss: 0.5734 - accuracy: 0.7031 - val_loss: 0.5642 - val_accuracy: 0.7333
Epoch 2/10
40/40 [==============================] - 5s 115ms/step - loss: 0.5568 - accuracy: 0.7328 - val_loss: 0.5648 - val_accuracy: 0.7267
Epoch 3/10
40/40 [==============================] - 5s 114ms/step - loss: 0.5535 - accuracy: 0.7484 - val_loss: 0.5489 - val_accuracy: 0.7633
Epoch 4/10
40/40 [==============================] - 5s 114ms/step - loss: 0.5337 - accuracy: 0.7422 - val_loss: 0.5357 - val_accuracy: 0.7567
Epoch 5/10
40/40 [==============================] - 5s 114ms/step - loss: 0.5233 - accuracy: 0.7594 - val_loss: 0.5301 - val_accuracy: 0.7533
Epoch 6/10
40/40 [==============================] - 5s 114ms/step - loss: 0.5342 - accuracy: 0.7477 - val_loss: 0.5277 - val_accuracy: 0.7667
Epoch 7/10
40/40 [==============================] - 5s 115ms/step - loss: 0.5219 - accuracy: 0.7563 - val_loss: 0.5210 - val_accuracy: 0.7567
Epoch 

one layer of conv2d has generate 896 params and the other hand one layer of separableconv2d has generate just 155 param (you can see above)
val accuracy of both architecture are all most same 